// @tolgee/react/dist/tolgee-react-server.esm.js
import React, { cache } from 'react';
export * from '@tolgee/web';

function unwrapSingleElementArray(value) {
    if (Array.isArray(value) && value.length === 1) {
        return value[0];
    }
    else {
        return value;
    }
}
const wrapTagHandlers = (params) => {
    if (!params) {
        return undefined;
    }
    const result = {};
    Object.entries(params || {}).forEach(([key, value]) => {
        if (typeof value === 'function') {
            result[key] = (chunk) => {
                return value(addReactKeys(chunk));
            };
        }
        else if (React.isValidElement(value)) {
            const el = value;
            result[key] = (chunk) => {
                return el.props.children === undefined && (chunk === null || chunk === void 0 ? void 0 : chunk.length)
                    ? React.cloneElement(el, {}, addReactKeys(chunk))
                    : React.cloneElement(el);
            };
        }
        else {
            result[key] = value;
        }
    });
    return result;
};
const addReactKeys = (children) => {
    const val = unwrapSingleElementArray(children);
    if (Array.isArray(val)) {
        return val.map((item, i) => (React.createElement(React.Fragment, { key: i }, item)));
    }
    else {
        return val;
    }
};

const TBase = (props) => {
    const key = props.keyName || props.children;
    if (key === undefined) {
        // eslint-disable-next-line no-console
        console.error('T component: keyName not defined');
    }
    const defaultValue = props.defaultValue ||
        (props.keyName ? props.children : undefined);
    const translation = addReactKeys(props.t({
        key: key,
        params: wrapTagHandlers(props.params),
        defaultValue,
        noWrap: props.noWrap,
        ns: props.ns,
        language: props.language,
    }));
    return React.createElement(React.Fragment, null, translation);
};

// @ts-ignore
const createServerInstance = ({ createTolgee, getLocale, }) => {
    const getTolgeeInstance = cache(async (locale) => {
        const tolgee = await createTolgee(locale);
        await tolgee.run();
        return tolgee;
    });
    const getTolgee = async () => {
        const locale = await getLocale();
        const tolgee = await getTolgeeInstance(locale);
        return tolgee;
    };
    const getTranslate = async () => {
        const tolgee = await getTolgee();
        return tolgee.t;
    };
    async function T(props) {
        const t = await getTranslate();
        return React.createElement(TBase, Object.assign({ t: t }, props));
    }
    return { getTolgeeInstance, getTolgee, getTranslate, T };
};

export { TBase, createServerInstance };
